runtime/src/accounts_db.rs

```diff
@@ -6511,6 +6511,51 @@ impl AccountsDb {
         self.flush_slot_cache_with_clean(&[slot], None::<&mut fn(&_, &_) -> bool>, None)
     }
 
+    /// 1.13 and some 1.14 could produce legal snapshots with more than 1 append vec per slot.
+    /// This is now illegal at runtime in the validator.
+    /// However, there is a clear path to be able to support this.
+    /// So, combine all accounts from 'slot_stores' into a new storage and return it.
+    /// This runs prior to the storages being put in AccountsDb.storage
+    pub(crate) fn combine_multiple_slots_into_one_at_startup(
+        path: &Path,
+        id: AppendVecId,
+        slot: Slot,
+        slot_stores: &HashMap<AppendVecId, Arc<AccountStorageEntry>>,
+    ) -> Arc<AccountStorageEntry> {
+        let size = slot_stores.values().map(|storage| storage.capacity()).sum();
+        let storage = AccountStorageEntry::new(path, slot, id, size);
+
+        // get unique accounts, most recent version by write_version
+        let mut accum = HashMap::<Pubkey, StoredAccountMeta<'_>>::default();
```

Comment:
Not sure if it matters, but I just realized, you can also use pointers, instead of copying public keys.
```suggestion
        let mut accum = HashMap::<&Pubkey, StoredAccountMeta<'_>>::default();
```
But it requires [a small change](https://github.com/solana-labs/solana/pull/30576), as currently, `StoredAccountMeta` returns references with an overly strict lifetimes.

Reply:
this runs into the borrow checker again. the key would be borrowed from the value.

Reply:
Yes, you need https://github.com/solana-labs/solana/pull/30576 as well.
I've checked it locally, before posting this time :)
`StoredAccountMeta` is overly conservative with the lifetimes it puts on the returned references.
The `Pubkey` comes from the `slot_stores`, which lives long enough.
`StoredAccountMeta` itself only holds references into the `slot_stores`.  And the compiler allows you to store it in the `accum` map just fine.
So, it is reasonable to expect that you should be able to store references to the `Pubkey`s as well, rather than having to copy them.
`Pubkey`s are 32 bytes, while the pointer is 8, so the saving is not that big.